hello-world/2023-02-20_testable_clock.md

```diff
@@ -0,0 +1,104 @@
+﻿# A testable clock
+Every developer that ever build an application that has any time dependency
+knows that testing such dependencies is not straight forward. The most pressing
+issue is arguably that most programming languages (including C#, PHP, Java, SQL)
+do not allow you to change the behaviour of the (system) clock/date time provider.
+So how to overcome this problem?
+
+## Stubable Date Time provider
+With the addoption of depdency injection/inverse of control, most programmers
+have developed a habbit of creating services for everything. DI frameworks do
+most of the work, and a (unit) test implementation helps while testing.
+
+``` C#
+public interface IClock
+{
+    DateTime UtcNow();
+}
+```
+
+This is a common way of solving the issue. By using an abstract class instead,
+we can easily define the shared clock logic in the base class, as most of it
+is the same for all implementations:
+
+``` C#
+public abstract class Clock
+{
+	public static readonly Clock System = new SystemClock();
+
+	public abstract DateTime UtcNow();
+	
+	public DateOnly Today() => DateOnly.FromDateTie(UtcNow());
+	
+	public DateOnly Yesterday() => Today().AddDays(-1);
+	
+	public override ToString() => UtcNow().ToString("U");
+
+    private sealed class SystemClock : Clock
+	{
+	    public override DateTime UtcNow() => DateTime.UtcNow;
+	}
+}
+
+public sealed class TestClock
```

Comment:
What's the purpose of the TestClock vs SystemClock here, should it inherit from Clock? And what can you do with the SystemClock?